wallet: add asset coin locking

guggero opened the pull request:
~Depends on https://github.com/lightninglabs/taproot-assets/pull/429.~

Fixes https://github.com/lightninglabs/taproot-assets/issues/261.

This PR introduces two new "coin locks":
 - One on the wallet level (similar to `btcwallet`'s `WithCoinSelectLock`) that allows multiple DB calls to be made while holding an exclusive lock (e.g. do coin selection then "lease" a specific coin to spend, those cannot be done in a single DB transaction the way the code is structured)
 - One on the `managed_utxo` level, which is called a "lease" (to reflect the terminology used in `lnd`: An application can "lease" (=lock/reserve) a bitcoin level UTXO to be spent in a transaction. So all assets committed to that BTC output are not available for other RPC users as long as the UTXO is being leased.

TODOs (might decide to put those into a second PR):
 - ~Allow releasing a lease (unlock coin) through RPC~
 - ~Allow filtering assets by `leased=true/false`~

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -202,6 +205,14 @@ type ActiveAssetsStore interface {
 
+	// LeaseManagedUTXO leases a managed UTXO identified by the passed
+	// serialized outpoint.
+	LeaseManagedUTXO(ctx context.Context, arg LeaseManagedUTXO) error
```
Same type and method name here? Maybe for the type name: `InsertLeasedUTXO`?

guggero replied:
Right, makes sense. I went with `UpdateUTXOLease` and `DeleteUTXOLease` since we'll also want to be able to extend the lease in the future.

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -202,6 +205,14 @@ type ActiveAssetsStore interface {
+	// serialized outpoint.
+	LeaseManagedUTXO(ctx context.Context, arg LeaseManagedUTXO) error
+
+	// ReleaseManagedUTXO releases a managed UTXO identified by the passed
```
A biit of am ambiguity here re: release vs re-lease. Maybe we should move to unlock/lock?

guggero replied:
I called it `DeleteUTXOLease` here but will try to make it clear that this unlocks a UTXO on the higher levels. Wanted to re-use as much of the terminology we use in `lnd` (even if it's not optimal) to avoid confusion.

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -358,6 +369,19 @@ type ChainAsset struct {
 	// anchor output.
 	AnchorTapscriptSibling []byte
+
+	// AnchorLeaseOwner is the identity of the application that currently
```
Not sure this needs to be part of the `ChainAsset` struct, but still getting through the diff.

Roasbeef replied:
Also perhaps we can constrain this a bit further, I think it's meant to be a `[32]byte`?

guggero replied:
I think it's useful to eventually show this in the asset list RPC. That way we don't need an explicit `ListAssetLeases` RPC, we can just add the `leased` boolean as a filter on the `ListAssets` request params.

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -718,6 +742,17 @@ func dbAssetsToChainAssets(dbAssets []ConfirmedAsset,
+		// valid and hasn't expired.
+		owner := sprout.AnchorLeaseOwner
+		expiry := sprout.AnchorLeaseExpiry
+		if len(owner) > 0 && expiry.Valid &&
```
How/when are leases to be cleaned up?

guggero replied:
This isn't currently done. Not sure what the best approach would be. In `btcwallet` I think we check the leases on certain actions. We could do the same here as well, e.g. whenever we list the coins or insert a new lease.
Or we could do it with a timed job. Going to see what makes most sense while addressing the other comments.

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -1425,10 +1469,14 @@ func queryChainAssets(ctx context.Context, q ActiveAssetsStore,
 	anchorPoint wire.OutPoint, groupKey *asset.GroupKey,
-	scriptKey *asset.ScriptKey) (*tapfreighter.AnchoredCommitment, error) {
+	scriptKey *asset.ScriptKey,
+	mustBeLeased bool) (*tapfreighter.AnchoredCommitment, error) {
```
So this is a sort of sanity check we want to implement in the constraints of a DB transaction? What about a similar coarse coin selection mutex to what we have in `lnd`? 

Looking at it from another angle: wouldn't one already fail a coin selection attempt if the outputs needed are already leased? Since first we'd try to perform coin selection with either `ListEligibleCoins` or `SelectForAmount`?

guggero replied:
The `FetchCommitment` is needed when we go to anchor a signed vPSBT. So at that point the caller MUST have leased the outpoint already.
Flow is:
 1. Fund vPSBT -> Calls `ListEligibleCoins`, then `SelectForAmount`, then adds lease for the "selected" coin. PSBT is returned to caller (or freighter state machine is advanced, but either way, this step is done)
 2. Sign vPSBT -> No DB interaction required
 3. AnchorVirtualPsbts -> Calls `FetchCommitment` to load the inputs required to create the new tree. At this point we select the coin being spent and we need to make sure it was previously leased or is currently still leased.

The other place where `FetchCommitment` is called is for creating an asset ownership proof. We also need to select the full commitment for that, but there we don't care if the asset is currently leased or not.

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -1468,12 +1523,65 @@ func (a *AssetStore) ListEligibleCoins(
+// LeaseCoin leases/locks/reserves a coin for the given lease owner until the
+// given expiry. This is used to prevent multiple concurrent coin selection
+// attempts from selecting the same coin.
+func (a *AssetStore) LeaseCoin(ctx context.Context, utxoOutpoint wire.OutPoint,
```
Hmm, I wonder if we want leasing at a lower level of abstractions: the full prev ID schema. I guess this works tho, since you can only ever spend that UTXO once. 

In the future, more granular leasing might be needed once we start to optimistically combine outgoing sends in the freighter.

guggero replied:
Makes sense. So we would either batch sends together in a timed manner. Or use RBF to replace a previously published anchoring TX? Either way, I think we'll need some more design around that and then probably go with a new column on the asset table itself.

Roasbeef commented on tapdb/sqlc/migrations/000002_assets.up.sql:
```diff
@@ -143,7 +143,19 @@ CREATE TABLE IF NOT EXISTS managed_utxos (
+    -- UTXO is being reserved/locked to be spent in an upcoming transaction and
+    -- that it should not be available for coin selection through any of the
+    -- wallet RPCs.
+    lease_owner BLOB,
```
Could alternatively be a distinct table, then the unlock/lock process is an insert/delete vs modification.

guggero replied:
I did look into that but found that it makes the coin selection SELECT statement quite a bit more complicated. Because then we'd need to do a `LEFT JOIN` on that leases table and also have WHERE statements that evaluates those columns based on the `leased` parameter.

Roasbeef replied:
Hmm not sure I follow, for lease aware coin selection, don't we just need to first get the set of leased asset IDs (the internal primary key one), then assert that the returned set isn't found amongst that table? So imagining it as a CTE. 

In either case will check out the latest posted version here.

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -320,6 +322,17 @@ type AssetStore struct {
+	// those methods must hold this coin lock. We use this in conjunction
+	// with an atomic  boolean, so we can check if the lock is held without
+	// having to acquire the mutex.
+	coinLock     sync.Mutex
```
I think this should live one level higher in the porter. So here: https://github.com/lightninglabs/taproot-assets/blob/edf59c62d5e4acd407d2675414e5d86e78eaf47b/tapfreighter/wallet.go#L143. 

In the past we've had some hard to track down bugs as a result of inserting user-space DB level mutexes. This way we also make minimal assumptions re concurrency control beyond what's provided by the DB.

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -320,6 +322,17 @@ type AssetStore struct {
+	// with an atomic  boolean, so we can check if the lock is held without
+	// having to acquire the mutex.
+	coinLock     sync.Mutex
+	coinLockHeld atomic.Bool
```
I don't think we'd need this atomic either. 

Similar to `lnd`, we'd also want to ensure that any relevant RPCs also respect the mutex. If they go to the wallet as suggested above, then that should be handled. Alternatively, they can grab the mutex itself, similar to the closure pattern we have in lnd.

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -2640,6 +2677,20 @@ func (a *AssetStore) FetchAssetMetaByHash(ctx context.Context,
+// WithCoinLock will acquire the coin lock and execute the passed function.
+// Callers to any coin related functions should use this to ensure that the
+// coin lock is held across the entire set of operations.
+func (a *AssetStore) WithCoinLock(fn func() error) error {
```
See comments above about moving this. I think sort of unclear with it being so low level exactly how the caller should interact with it.

Roasbeef commented on tapfreighter/interface.go:
```diff
@@ -87,8 +87,16 @@ type CoinLister interface {
+	// function. Callers to any coin related functions should use this to
+	// ensure that the coin lock is held across the entire set of
+	// operations.
+	WithCoinLock(fn func() error) error
```
I think this placement makes sense, but then I think the implementation can just reside within `CoinSelect ` (the concrete implementation).

Roasbeef commented on tapfreighter/wallet.go:
```diff
@@ -385,23 +394,34 @@ func (f *AssetWallet) FundPacket(ctx context.Context,
-	log.Infof("Identified %v eligible asset inputs for send of %d to %x",
-		len(eligibleCommitments), fundDesc.Amount, fundDesc.ID[:])
+	var selectedCommitments []*AnchoredCommitment
+	err := f.cfg.CoinSelector.WithCoinLock(func() error {
```
👍

Roasbeef commented on tapfreighter/wallet.go:
```diff
@@ -385,23 +394,34 @@ func (f *AssetWallet) FundPacket(ctx context.Context,
 
-	log.Infof("Identified %v eligible asset inputs for send of %d to %x",
-		len(eligibleCommitments), fundDesc.Amount, fundDesc.ID[:])
+	var selectedCommitments []*AnchoredCommitment
```
Perhaps we also want to have a wrapper type here that makes it easier to release all the coin in a single go?

Roasbeef commented on tapfreighter/wallet.go:
```diff
@@ -385,23 +394,34 @@ func (f *AssetWallet) FundPacket(ctx context.Context,
+		log.Infof("Identified %v eligible asset inputs for send of %d "+
+			"to %x", len(eligible), fundDesc.Amount, fundDesc.ID[:])
+
+		selectedCommitments, err = f.cfg.CoinSelector.SelectForAmount(
```
Looks like leasing the coins is missing here? Otherwise two callers can come in series, then obtain the same set of coins, and attempt to lease them both.

guggero replied:
Yes, that's added in a later commit. Added a TODO here for now to help with the review flow.

Roasbeef commented on tapdb/sqlc/queries/assets.sql:
```diff
@@ -676,6 +685,16 @@ WHERE (
 
+-- name: LeaseManagedUTXO :exec
+UPDATE managed_utxos
+SET lease_owner = @lease_owner, lease_expiry = @lease_expiry
```
See above about making this an insert. That way, we can have the unique key be the outpoint itself. Then if two callers try to lease the same UTXO, they get an error. Otherwise, with the current API the same coin can be leased twice.

guggero replied:
See comment above. I think it makes things unnecessarily hard.
And a coin cannot be leased twice because of the wallet-level coin lock:
 1. Caller A comes in, does: `WalletLock( ListEligibleCoins => returns coins X, Y, Z => SelectForAmount => select coin X => LeaseCoin(X))`
 2. Caller B comes in while call A is still ongoing. Once it acquires the lock, it does: `WalletLock( ListEligibleCoins => returns coins Y, Z => SelectForAmount => select coin Y => LeaseCoin(Y))`

Roasbeef replied:
Ah yeh that's indeed the case with the full API path as exists today. 

I think I was just viewing it more from the PoV of the lowest level data structure here: the table itself. If we used the `outpoint` (or `outpoint || assetID` or w/e) as a primary key, then we'd be able to entirely rely on the DB to preserve the invariant that something can only be leased once: another insert would fail. As is, it would be possible for to callers (strictly on the db level just to zoom in) to each lease the same UTXO, with the second replacing/mutating the `lease_owner` field. 

As you mention above it still works as is, but if we wanted to try and normalize things further, I think we'd put the leases in a distinct table. Though up until now, things were normalized mostly on a best effort basis, so we don't necessarily have a strict policy here. Either way, just wanted to point this out as we recently-ish learned that the `sqlite` migration support isn't as fully featured as `postgres` so it'll be more difficult to change later.

Roasbeef commented on tapdb/assets_store_test.go:
```diff
@@ -677,16 +683,21 @@ func TestSelectCommitment(t *testing.T) {
-				ctx, tc.constraints,
-			)
+			var selectedAssets []*tapfreighter.AnchoredCommitment
+			err := store.WithCoinLock(func() error {
```
Re the prior comments, we then wouldn't have to leak the details of the coin selection lock into the database level. Instead, we can have a higher level unit test that just gives a mocked implementation of `CoinLister`.

Roasbeef commented on tapfreighter/chain_porter.go:
```diff
@@ -202,14 +202,10 @@ func (p *ChainPorter) assetsPorter() {
+			// possible in its own goroutine. The status will be
+			// reported through the different channels of the send
+			// package.
+			go p.advanceState(sendPkg, req.kit())
```
👍

Roasbeef commented on tapfreighter/wallet.go:
```diff
@@ -152,6 +174,27 @@ func (s *CoinSelect) ListEligibleCoins(ctx context.Context,
+func (s *CoinSelect) LeaseCoin(ctx context.Context, utxoOutpoint wire.OutPoint,
+	leaseOwner []byte, expiry time.Time) error {
+
+	return s.coinLister.LeaseCoin(ctx, utxoOutpoint, leaseOwner, expiry)
```
Not related to this line of code, but wanting to jot this down: I think we should make sure that once an outgoing package has been written to disk, with the transfer transaction broadcast (but not yet confirmed), that UTXO shouldn't be available for leasing. Otherwise, if confirmations take a while, then we end up double spending ourselves. 

With the way things are in `btcwallet`, once a mined transaction is inserted, then the lease is also removed: https://github.com/btcsuite/btcwallet/blob/07be54bc22cf4dd2dd16a9788d2fc88952c438ae/wtxmgr/tx.go#L462-L471

Roasbeef commented on tapfreighter/wallet.go:
```diff
@@ -416,6 +459,21 @@ func (f *AssetWallet) FundPacket(ctx context.Context,
 
+		// We now need to lock/lease/reserve those selected coins so
+		// that they can't be used by other processes.
+		expiry := time.Now().Add(defaultCoinLeaseDuration)
```
👍

Roasbeef reviewed (changes requested):
Great work! Changes look to be headed in the right direction at a high level. I think my major review comment relates to exactly _where_ the coin selection lock should live it. Right now it lives at the bottom of the abstraction layer in the db, which also means we now have a mutex in the way of db code. In the past (eg lnd), we've run into issues with user space mutexes on top of what the db is already doing. I think we should try to lift the mutex up to the highest level, understanding that some calls may just want a snapshot of the db, and don't necessarily need the same level of contextual locking as needed when trying to perform a transfer. 

Also as is, it looks like an output can be locked twice. If the lease are a new table, and then we use an outpoint as the primary key, we can reject such behavior as the unique conflict then fails. Finally, I didn't see where the leases are expired, seems to be sort of implicit rn?

ffranr commented on tapdb/sqlc/queries/assets.sql:
```diff
@@ -398,7 +400,14 @@ JOIN internal_keys
+           WHEN sqlc.narg('leased') = true THEN
+               (utxos.lease_owner IS NOT NULL AND utxos.lease_expiry > @now)
+           WHEN sqlc.narg('leased') = false THEN
+               (utxos.lease_owner IS NULL OR utxos.lease_expiry <= @now)
```
Perhaps we should add `OR utxos.lease_expiry IS NULL` to this `WHEN` clause.

guggero replied:
Good catch, added.

guggero commented:
Latest push addresses most comments. Though I think I should address the following TODOs before a full re-review:
 - Add more unit tests around the leases and the `CoinSelect` struct
 - Periodically (or upon certain requests) remove old leases
 - Allow filtering assets by lease status on the RPC, assert leases in itest

ffranr reviewed (commented):
Nice work! Looks good to me. I don't have too much to add beyond Laolu's comments.

I think we should probably add an integration test to ensure that the send process fails if all candidate UTXOs are locked.

guggero commented:
> I think we should probably add an integration test to ensure that the send process fails if all candidate UTXOs are locked.

We do that here: https://github.com/lightninglabs/taproot-assets/blob/814ea4b374179fda9a119629e0602b333c749563/itest/send_test.go#L766

Roasbeef commented on tapdb/assets_store.go:
```diff
@@ -726,7 +764,12 @@ func dbAssetsToChainAssets(dbAssets []ConfirmedAsset,
-	var assetFilter QueryAssetFilters
+	assetFilter := QueryAssetFilters{
+		Now: sql.NullTime{
+			Time:  time.Now().UTC(),
```
Not blocking at all here, but I think we'll want to eventually start to use `clock.Clock` (from lnd) here so we can eventually mock out the time for unit tests.

Roasbeef commented on tapfreighter/chain_porter.go:
```diff
@@ -24,10 +24,6 @@ import (
 type ChainPorterConfig struct {
-	// CoinSelector is the interface used to select input coins (assets)
-	// for the transfer.
-	CoinSelector CoinSelector
```
👍

Roasbeef commented on tapfreighter/wallet.go:
```diff
@@ -383,25 +412,15 @@ func (f *AssetWallet) FundPacket(ctx context.Context,
+		MinAmt:   fundDesc.Amount,
 	}
-
-	log.Infof("Identified %v eligible asset inputs for send of %d to %x",
```
Nice simplification!

Roasbeef reviewed (commented):
The change set is looking pretty good! Will pull it down and mess around with some of the new functionality on the CLI to see how intuitive/surprising it is

guggero commented:
Added all unit tests that I wanted to add, ready for another round of review.

ffranr commented on tapdb/assets_store.go:
```diff
@@ -722,15 +749,33 @@ func dbAssetsToChainAssets(dbAssets []ConfirmedAsset,
+		owner := sprout.AnchorLeaseOwner
+		expiry := sprout.AnchorLeaseExpiry
+		if len(owner) > 0 && expiry.Valid &&
+			expiry.Time.UTC().After(a.clock.Now().UTC()) {
```
nit: the if condition is getting complicated here.

I would consider evaluating a bool variable before the if clause:
```
var (
    hasOwner = len(owner) > 0
    isExpired = expiry.Valid && expiry.Time.UTC().After(a.clock.Now().UTC())
)

if hasOwner && isExpired {
```

Or maybe not, it's a style.

ffranr commented on tapdb/assets_store.go:
```diff
@@ -1819,9 +1819,13 @@ func (a *AssetStore) queryCommitments(ctx context.Context,
+// UTXOs for forever, which means the expiry should be far in the future.
 func (a *AssetStore) LogPendingParcel(ctx context.Context,
-	spend *tapfreighter.OutboundParcel) error {
+	spend *tapfreighter.OutboundParcel, finalLeaseOwner [32]byte,
+	finalLeaseExpiry time.Time) error {
```
nit: I don't understand why we need the `final` prefix here in `finalLeaseOwner` and `finalLeaseExpiry`. Consider removing?

ffranr commented on tapdb/assets_store_test.go:
```diff
@@ -594,6 +614,246 @@ func (a *assetGenerator) bindKeyGroup(i int, op wire.OutPoint) *btcec.PublicKey
+	require.NoError(t, err)
+	require.Len(t, selectedAssets, numAssets)
+
+	// Now we lease the first asset for 1 hour. This will cause the second
+	// one also to be leased, since it's on the same anchor transaction.
```
:+1:

ffranr reviewed (approved):
I think I've managed to review the changes since my last review. Everything looks good to me. Nice addition of the clock to different db stores. And a test which ensures that locking an outpoint affects each asset anchored at that outpoint :+1:

Roasbeef reviewed (approved):
LGTM 🦆
